server/common/config/app_config.py

```diff
@@ -0,0 +1,172 @@
+import yaml
+from flatten_dict import unflatten
+import os
+from server.common.config.dataset_config import DatasetConfig
+from server.common.config.server_config import ServerConfig
+from server.common.errors import ConfigurationError
+
+
+class AppConfig(object):
+    """AppConfig stores all the configuration for cellxgene.  The configuration is divided into two main parts:
+    server attributes, and dataset attributes.
+    The server_config contains attributes that refer to the server process as a whole.
+    The default_dataset_config refers to attributes that are associated with the features and
+    presentations of a dataset.
+    The dataset config attributes can be overridden depending on the url by which the
+    dataset was accessed.  These are stored in dataroot_config.
+    AppConfig has methods to initialize, modify, and access the configuration.
+    """
+
+    def __init__(self):
+
+        # the default configuration (see default_config.py)
+        # TODO @madison -- if we always read from the default config (hard coded path) can we set those values as
+        #  defaults within the config class?
+        self.default_config = self.get_default_config()
+        # the server configuration
+        self.server_config = ServerConfig(self, self.default_config["server"])
+        # the dataset config, unless overridden by an entry in dataroot_config
+        self.default_dataset_config = DatasetConfig(None, self, self.default_config["dataset"])
+        # a dictionary of keys to DatasetConfig objects.  Each key must exist in the multi_dataset__dataroot
+        # attribute of the server_config.
+        self.dataroot_config = {}
+
+        # Set to true when config_completed is called
+        self.is_completed = False
+
+    @staticmethod
+    def get_default_config():
+        try:
+            with open(f"{os.getenv('PROJECT_ROOT')}/default_config.yml", 'r') as default_config:
+                return yaml.safe_load(default_config)
+        except yaml.YAMLError as e:
```

Comment:
Maybe also toss in a `except OSError as err:` to catch the case that the file doesn't exist and explain that it wasn't able to find the file?